src/main/kotlin/com/ajax/cryptocurrency/service/CryptocurrencyService.kt

```diff
@@ -0,0 +1,65 @@
+package com.ajax.cryptocurrency.service
+
+import com.ajax.cryptocurrency.exception.CryptocurrencyPriceNotFoundException
+import com.ajax.cryptocurrency.exception.InvalidParamsForThePageException
+import com.ajax.cryptocurrency.model.Cryptocurrency
+import com.ajax.cryptocurrency.repository.CryptocurrencyRepository
+import org.springframework.data.domain.PageRequest
+import org.springframework.data.domain.Sort
+import org.springframework.stereotype.Service
+import java.io.File
+import java.io.FileWriter
+
+@Service
+class CryptocurrencyService(private val cryptocurrencyRepository: CryptocurrencyRepository) {
+    fun findMinMaxPriceByCryptocurrencyName(name: String, sortOrder: Int?): Cryptocurrency {
+        val unwrappedSortOrder = sortOrder ?: throw IllegalArgumentException("Sort order cannot be null")
+
+        val priceMinMax = cryptocurrencyRepository.findMinMaxByName(name, unwrappedSortOrder)
+            ?: throw CryptocurrencyPriceNotFoundException("Price of $name cryptocurrency not found")
+
+        return priceMinMax
+    }
+
+    fun findAll(): List<Cryptocurrency> {
+        return cryptocurrencyRepository.findAll()
+    }
+
+    fun getCryptocurrencyPages(name: String?, pageNo: Int, pageSize: Int): MutableList<Cryptocurrency?> {
+        when {
+            cryptocurrencyRepository.findAll()
+                .isEmpty() -> throw CryptocurrencyPriceNotFoundException("cryptocurrency prices not found")
+
+            pageNo < 0 || pageSize < 0 -> throw InvalidParamsForThePageException("invalid params for the page")
```

Comment:
Should be validated at controller layer.

It could be done by:
1. Annotate a contorller with @Validated
2. `page` and `size` annotate with `@Min(1)`. 
     Yes, min value for both should be: `1`. Page counting should be started with 1. Requested page size must be greater than `0`

Please, don't forget to check whether it works

Reply:
fixed